x/ecocredit/server/marketplace/buy.go

```diff
@@ -0,0 +1,255 @@
+		} else {
+			// if we can partial fill, we just delete the sellOrder and take whatever
+			// credits are left from that order.
+			if err := k.stateStore.SellOrderTable().Delete(ctx, sellOrder); err != nil {
+				return err
+			}
+			purchaseQty = sellOrderQty
+		}
+	}
+	if newSellOrderQty.IsZero() { // remove the sell order if no credits are left
+		if err := k.stateStore.SellOrderTable().Delete(ctx, sellOrder); err != nil {
+			return err
+		}
+	} else { // update the sell order with the new value otherwise
+		sellOrder.Quantity = newSellOrderQty.String()
+		if err = k.stateStore.SellOrderTable().Update(ctx, sellOrder); err != nil {
+			return err
+		}
+	}
+
+	// remove the credits from the seller's escrowed balance
+	sellerBal, err := k.coreStore.BatchBalanceTable().Get(ctx, sellOrder.Seller, sellOrder.BatchId)
+	if err != nil {
+		return err
+	}
+	escrowBal, err := math.NewDecFromString(sellerBal.Escrowed)
+	if err != nil {
+		return err
+	}
+	escrowBal, err = math.SafeSubBalance(escrowBal, purchaseQty)
+	if err != nil {
+		return err
+	}
+	sellerBal.Escrowed = escrowBal.String()
+	if err = k.coreStore.BatchBalanceTable().Update(ctx, sellerBal); err != nil {
+		return err
+	}
+
+	// update the buyers balance and the batch supply
+	supply, err := k.coreStore.BatchSupplyTable().Get(ctx, sellOrder.BatchId)
+	if err != nil {
+		return err
+	}
+	buyerBal, err := k.coreStore.BatchBalanceTable().Get(ctx, buyerAcc, sellOrder.BatchId)
+	if err != nil {
+		if ormerrors.IsNotFound(err) {
+			buyerBal = &ecocreditv1.BatchBalance{
+				Address:  buyerAcc,
+				BatchId:  sellOrder.BatchId,
+				Tradable: "0",
+				Retired:  "0",
+				Escrowed: "0",
+			}
+		} else {
+			return err
+		}
+	}
+	if sellOrder.DisableAutoRetire { // if auto retire is disabled, we move the credits into the buyer's/supply's tradable balance.
+		tradableBalance, err := math.NewDecFromString(buyerBal.Tradable)
+		if err != nil {
+			return err
+		}
+		tradableBalance, err = math.SafeAddBalance(tradableBalance, purchaseQty)
+		if err != nil {
+			return err
+		}
+		buyerBal.Tradable = tradableBalance.String()
+
+		supplyTradable, err := math.NewDecFromString(supply.TradableAmount)
+		if err != nil {
+			return err
+		}
+		supplyTradable, err = math.SafeAddBalance(supplyTradable, purchaseQty)
+		if err != nil {
+			return err
+		}
+		supply.TradableAmount = supplyTradable.String()
+	} else {
+		retiredBalance, err := math.NewDecFromString(buyerBal.Retired)
+		if err != nil {
+			return err
+		}
+		retiredBalance, err = math.SafeAddBalance(retiredBalance, purchaseQty)
+		if err != nil {
+			return err
+		}
+		buyerBal.Retired = retiredBalance.String()
+
+		supplyRetired, err := math.NewDecFromString(supply.RetiredAmount)
+		if err != nil {
+			return err
+		}
+		supplyRetired, err = math.SafeAddBalance(supplyRetired, purchaseQty)
+		if err != nil {
+			return err
+		}
+		supply.RetiredAmount = supplyRetired.String()
+	}
```

Comment:
lot of repeated "decode + add". Would be good to create a helper function for that

Reply:
BTW, why did we decide to store numbers as string rather than bytes?

Reply:
not sure, just how its been done so far. perhaps we need an issue/discussion on it

Reply:
Yes, let's continue in an issue - it's the same discussion as we have for addresses.  Just for note that ledger eventually will use different serialization anyway.